### PR TITLE
[E112-MG < T0851-MG] Enforce schema on vertex creation

### PR DESCRIPTION
**Pull request by jbajic**
[master < Task] PR
- [ ] ~~Check, and update documentation if necessary~~
- [ ] ~~Update [changelog](https://docs.memgraph.com/memgraph/changelog)~~
- [x] Provide the full content or a guide for the final git message

This PR introduces a breaking change of enforcing schema validation upon vertex manipulation:
- Vertex creation,
- Vertex properties update and
- Vertex adding/removing labels
- Major changes to unit tests
Also, the following features have been excluded from vertex validation:
- `mg_import`
- query modules
- replication

This PR won't fix any e2e tests, integration tests, or mgbench tests.

Commit Message:
- Separating schema definition from schema validation
- Updating `vertex_accessor` and `db_accessors` with necessary methods
- Adding a primary label to `Vertex`
- Adding schema tests
- Updating existing tests for storage v3, and deprecating old:
  -  `interpreter` => `interpreter_v2`
  -  `query_plan_accumulate_aggregate` => `storage_v3_query_plan_accumulate_aggregate`
  - `query_plan_create_set_remove_delete` => `storage_v3_query_plan_create_set_remove_delete`
  - `query_plan_bag_semantics` => `storage_v3_query_plan_bag_semantics`
  - `query_plan_edge_cases` => `storage_v3_query_plan_edge_cases`
  - `query_plan_v2_create_set_remove_delete` => `storage_v3_query_plan_v2_create_set_remove_delete`
  - `query_plan_match_filter_return` => `storage_v3_query_plan_match_filter_return`

### CODE REVIEW

File: src/storage/v2/vertex_accessor.cpp
Comment (antaljanosbenjamin) [inline src/storage/v2/vertex_accessor.cpp:124]:
```
    has_label = VertexHasLabel(vertex_, label);
```

File: src/storage/v2/vertex_accessor.cpp
Comment (antaljanosbenjamin) [inline src/storage/v2/vertex_accessor.cpp:164]: After checking the PR, two questions here:
1. Do we need to expose the priamry label of the vertex? I think until we don't use it, we can emit this function.
2. I think I forgot about the existing and deleted stuff. So if we want to expose the primary label, then maybe it worth to check whether the vertex exists/deleted.

File: src/storage/v2/schema_validator.cpp
Comment (antaljanosbenjamin) [inline src/storage/v2/schema_validator.cpp:46]: As `Schemas::GetSchema` is called in the hot path and it contains a vector, consider using a pointer instead. If the function returns a `nullptr`, it means the schema doesn't exists. This way you can avoid copying the vector inside the schema.

File: src/storage/v2/schema_validator.cpp
Comment (antaljanosbenjamin) [inline src/storage/v2/schema_validator.cpp:66]:
```
      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_HAS_NO_PRIMARY_PROPERTY, primary_label, schema_type);
```

or 

```
      return SchemaViolation(SchemaViolation::ValidationStatus::SCHEMA_HAS_NO_PROPERTY, primary_label, schema_type);
```

Would be a better name in my opinion.

File: src/storage/v2/schema_validator.cpp
Comment (antaljanosbenjamin) [inline src/storage/v2/schema_validator.cpp:86]: This should be an assert in my opinion. This function shouldn't be called with non existing primary label, it is only valid for `ValidateCreateVertex`.

File: src/storage/v2/schema_validator.cpp
Comment (antaljanosbenjamin) [inline src/storage/v2/schema_validator.cpp:102]:
```suggestion
    return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_MODIFY_PRIMARY_LABEL, label);
```

The name `VERTEX_ALREADY_HAS_PRIMARY_LABEL` suggest for me that a vertex might exist without a primary label, which shouldn't be allowed in my opinion. Also this function is also called from `RemoveLabel`, which doesn't want to add any new label.

File: tests/unit/interpreter_v2.cpp
Comment (antaljanosbenjamin) [inline tests/unit/interpreter_v2.cpp:1]: When and how do you plan to migrate these changes to query-v2/storage-v3? I know we talked about it, but forgot the conclusion.

File: tests/unit/storage_v3_query_plan_accumulate_aggregate.cpp
Comment (antaljanosbenjamin) [inline tests/unit/storage_v3_query_plan_accumulate_aggregate.cpp:1]: Could you please explain why these tests has `storage_v3` in their name but using `storage-v2`?

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

File: src/query/db_accessor.hpp
Comment (jbajic) [reply src/query/db_accessor.hpp:122]: It is not a problem at all but my question is why not keep this validation here, isn't it better than t the validation happens as soon as possible?

File: src/query/db_accessor.hpp
Comment (jbajic) [reply src/query/db_accessor.hpp:18]: I will check includes in other files as well then, add this info to our c++ convention page

File: src/query/common.hpp
Comment (jbajic) [reply src/query/common.hpp:127]: This was something I was trying out, but it is not needed anymore, will remove it

File: tests/unit/storage_v3_query_plan_accumulate_aggregate.cpp
Comment (jbajic) [reply tests/unit/storage_v3_query_plan_accumulate_aggregate.cpp:1]: Because the schema is currently in storage-v2 and not in v3, and that will be all moved in the next PR, both query and storage to their next version, since the schema epic began before storage-v3 was introduced

File: tests/unit/interpreter_v2.cpp
Comment (jbajic) [reply tests/unit/interpreter_v2.cpp:1]: In the PR after we merge this one

File: src/query/db_accessor.hpp
Comment (jbajic) [reply src/query/db_accessor.hpp:122]: This was resolved in conversation between Benjamin and me, and the conclusion is we want to keep this logic in the storage engine for now because of these reasons:
 - query engine might send data that are not part of the schema defined  in the storage engine, because of some bug  especially in the beggining
 - validating in the query engine can be considered and optimization that will be possibly implemented in the later stages

File: tests/unit/storage_v3_schema.cpp
Comment (antaljanosbenjamin) [inline tests/unit/storage_v3_schema.cpp:42]: I am wondering whether we need this storage at all. I think it is not necessary. If you want to generate the property/label ids, then I think a NameIdMapper + two helper function are enough. The reason why I think this can be beneficial is it makes harder to introduce any dependency on the storage itself within these tests.

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

File: src/storage/v2/schema_validator.cpp
Comment (jbajic) [reply src/storage/v2/schema_validator.cpp:66]: I will for the first one, since all status already have vertex suffix

File: tests/unit/CMakeLists.txt
Comment (spacejam) [inline tests/unit/CMakeLists.txt:309]: does it make more sense to remove these, rather than have them commented out?

File: src/storage/v2/vertex_accessor.cpp
Comment (jbajic) [reply src/storage/v2/vertex_accessor.cpp:164]: 1. This is used throughout tests to confirm that the correct `PrimaryLabel` is set
2. I agree I will add that logic for fetching `PrimaryLabel`

File: tests/unit/interpreter_v2.cpp
Comment (jbajic) [reply tests/unit/interpreter_v2.cpp:138]: I forgot about it, when fixing tets I've commented out unrelated parts to spped up

File: tests/unit/storage_v3_query_plan_bag_semantics.cpp
Comment (jbajic) [reply tests/unit/storage_v3_query_plan_bag_semantics.cpp:35]: That makes sense

File: tests/unit/storage_v3_query_plan_create_set_remove_delete.cpp
Comment (jbajic) [reply tests/unit/storage_v3_query_plan_create_set_remove_delete.cpp:484]: You are right

File: tests/unit/storage_v3_schema.cpp
Comment (jbajic) [reply tests/unit/storage_v3_schema.cpp:40]: Yes will move this to initialization

File: tests/unit/CMakeLists.txt
Comment (jbajic) [reply tests/unit/CMakeLists.txt:309]: Hmm I don't see why, unless we want to explicitly track for some time which tests we have moved and which we haven't

Comment (jbajic) [reply tests/unit/CMakeLists.txt:309]: What do you say @antaljanosbenjamin?

File: tests/unit/query_v2_query_plan_match_filter_return.cpp
Comment (antaljanosbenjamin) [inline tests/unit/query_v2_query_plan_match_filter_return.cpp:1938]: Please replace this with `InsertVertexAndValidate`.

File: tests/unit/CMakeLists.txt
Comment (antaljanosbenjamin) [reply tests/unit/CMakeLists.txt:309]: > unless we want to explicitly track for some time which tests we have moved and which we haven't

How does this connected to this? I don't see it.

In general I think we should barely touch things that are on master, because we will constantly merging master to project pineapples. So maybe it make sense revert all the changes in v2 to leave it untouched.

File: src/storage/v2/durability/snapshot.cpp
Comment (jbajic) [reply src/storage/v2/durability/snapshot.cpp:716]: I didn't know there was a style guideline for TODOs

File: tests/unit/query_v2_query_plan_match_filter_return.cpp
Comment (jbajic) [reply tests/unit/query_v2_query_plan_match_filter_return.cpp:1938]: I've missed that sorry

File: tests/unit/CMakeLists.txt
Comment (jbajic) [reply tests/unit/CMakeLists.txt:309]: But we don't seperate tests, and we agreed that the new tests would have prefix storage_v3 or query_v2, but we would keep the old tests as well, the issues is that we cannot leave them running since they will be always failing. So we can either comment them out and always take a quick look to see which ones we had to convert or just delete them, is that clear?

File: src/storage/v2/vertex_accessor.hpp
Comment (jbajic) [reply src/storage/v2/vertex_accessor.hpp:35]: Currently, it is not possible due to `VertexAccessot` being copied, and if we have a `VertexValidator` that holds a reference object it cannot be copied, take a look at void `LabelIndex::Iterable::Iterator::AdvanceUntilValid() {` in `indices.cpp`. Do you think its worth doing something about it?

File: src/storage/v2/durability/snapshot.cpp
Comment (antaljanosbenjamin) [reply src/storage/v2/durability/snapshot.cpp:716]: I think even clang-tidy warns on this.

File: src/storage/v2/vertex_accessor.hpp
Comment (antaljanosbenjamin) [reply src/storage/v2/vertex_accessor.hpp]: I meant only as a constructor parameter, not as a member. So take a reference int the constructor, and then store the address of the object in a member variable. That way I think:
1. it is much more clear for the outside world that they cannot pass into nullptr
2. maybe it also gives a strong hint for the developer that they don't have to deal with nullptr in the member variable.

As I am checking it out, `SchemaValiator` can be passed as `const&`, because all of the interesting member functions are const qualified.

Comment (jbajic) [reply src/storage/v2/vertex_accessor.hpp]: Aha ok that makes sense, that's why you only marked the constructor, sorry for the misunderstanding

**Review by spacejam — APPROVED** [review]
AFAICT this looks good, but I defer to @antaljanosbenjamin wrt the `tests/unit/CMakeLists.txt` structuring feedback which is the main thing that caught my eye

File: tests/unit/CMakeLists.txt
Comment (jbajic) [reply tests/unit/CMakeLists.txt:309]: This will be handled in the next PR which moves all the changes do appropriate versions of storage and query engine

**Review by antaljanosbenjamin — APPROVED** [review]
LGTM.

File: src/storage/v2/vertex_accessor.hpp
Comment (antaljanosbenjamin) [reply src/storage/v2/vertex_accessor.hpp:47]: It isn't, because the `VertexAccessor` sometimes instantiated with a `Vertex` pointer that is `nullptr`.

**Comment by jbajic** [comment]
Next PR will move all the changes to query v2 and storage v3!

**Pull request merged**
